Virial stress

kinanbezem7 opened the pull request:
Added virial stress to particles.hpp and the HDF5 output for the pmb model. 
Added logic for calculating virial stress in compute full energy in force_pmb.hpp. 
Compiles and runs kalthoff winkler, however the virial stress in the output is zero. 

I think the forces may not be being calculated properly in the energy logic. I'm also unsure of how to handle bond breakage here, but my entire approach may be wrong so this might be totally unnecessary

streeve commented on bfg.jar:
Remove :)

streeve commented on src/force/CabanaPD_Force_PMB.hpp:
```diff
@@ -161,6 +161,43 @@ class Force<ExecutionSpace, ForceModel<PMB, Elastic, ModelParams...>>
+            //f( i, 1 ) += fy_i;
+            //f( i, 2 ) += fz_i;
+
+            auto virial_stress = particles.sliceVirialStress();
```
You are allowed to slice in kernel, but for performance we normally do it outside (as with volume)

kinanbezem7 replied:
Moved it outside with the volume

streeve commented on src/force/CabanaPD_Force_PMB.hpp:
```diff
@@ -161,6 +161,43 @@ class Force<ExecutionSpace, ForceModel<PMB, Elastic, ModelParams...>>
+
+            
+            // Calculate forces for virial stress
+            double fx_i = 0.0;
```
I had expected to use the current force, rather than recomputing - at least then you don't have to worry about them being zero. Unless I'm missing something it should just be the same way we currently interact with the volumes

kinanbezem7 replied:
Ah I misunderstood you before, I thought we had to recalculate it. I've tried to slice the force outside of the kernel and in order to do that I had to create a const version of sliceForce which I have added to particles.hpp. 
In turn I have changes the internal logic of the virialstress calculations but it is still zero in the output of kalthoff winkler even though the force output has values.

streeve commented on src/force/CabanaPD_Force_PMB.hpp:
```diff
@@ -161,6 +161,43 @@ class Force<ExecutionSpace, ForceModel<PMB, Elastic, ModelParams...>>
+            getDistanceComponents( x, u, i, j, xi, r, s, rx, ry, rz );
+
+            // Check if the bond is broken
+            //if ( model.criticalStretch( i, j, r, xi ) && !nofail( i ) && !nofail( j ) )
```
This version of the kernel doesn't include fracture, so we can remove this for now. Once it's working we can add below (where you see `Fracture`)

kinanbezem7 replied:
removed the bond breakage check from the virial stress logic as we are not recalculating the forces anymore

streeve reviewed (changes requested):
First review - let's get it working without fracture first

streeve commented on src/CabanaPD_Particles.hpp:
```diff
@@ -106,7 +106,7 @@ class Particles<MemorySpace, PMB, TemperatureIndependent, Dimension>
     // v, W, rho, damage,  type.
     using other_types =
-        Cabana::MemberTypes<double[dim], double, double, double, int>;
+        Cabana::MemberTypes<double[dim * 2],double[dim], double, double, double, int>; //added double[dim * 2] for virial stress
```
My bad, this should be dim*(dim+1)/2

streeve commented on src/force/CabanaPD_Force_PMB.hpp:
```diff
@@ -138,21 +138,25 @@ class Force<ExecutionSpace, ForceModel<PMB, Elastic, ModelParams...>>
     template <class PosType, class WType, class ParticleType,
               class NeighListType, class ParallelType>
     double computeEnergyFull( WType& W, const PosType& x, const PosType& u,
-                              const ParticleType& particles,
```
This was the first issue causing zeros

streeve reviewed (commented):
Zeros are because the only implemented stress is for Elastic (no fracture) with PMB, which requires modifying the elastic wave problem to use PMB. Looks good on my end making that change. Let me know if you have questions about adding this for fracture

kinanbezem7 commented:
I've made the suggested changes and it is now working with non-fracture problems using the PMB method. (I may have pushed an altered version of elastic wave problem that I used to test it...please disregard this). Will move on to implementing virial stress with fracture.

streeve commented:
> I've made the suggested changes and it is now working with non-fracture problems using the PMB method. (I may have pushed an altered version of elastic wave problem that I used to test it...please disregard this). Will move on to implementing virial stress with fracture.

Sounds good. One thing that will help quite a bit is using `pre-commit`, see: https://github.com/ORNL/CabanaPD/blob/main/CONTRIBUTING.md It will basically stop you from committing certain unintended things (unformatted code, merge conflicts, etc.)

One other note - we'll need to remove the comments tracking incremental changes before we merge (the fact that the force slice was moved outside the kernel is tracked in the git history and would likely be more confusing than not in the future, for example)

streeve commented:
> [1-s2.0-S0020768322000427-main.pdf](https://github.com/user-attachments/files/17478342/1-s2.0-S0020768322000427-main.pdf)
> 
> Added virial stress to pmb with fracture. Could you please doublecheck the volume summation. it should be normalised by the total volume of the horizon regardless of the number of points within the horizon. I'm thinking we might need to adjust the non-fracture volume too as it does not include vol(i) in the division.

@pabloseleson can you take a look?

streeve reviewed (changes requested):
Should we add this for LPS as well? Or wait for now?

pabloseleson commented:
Which equation is being implemented exactly?

kinanbezem7 commented:
> Which equation is being implemented exactly?

Eq. 29.

pabloseleson commented:
@kinanbezem7 : I looked at Eq. (29) in the reference [here](https://doi.org/10.1016/j.ijsolstr.2022.111478). There is something awkward in that equation. The RHS of the equation is for the stress tensor at X. However, the LHS of the equation doesn't depend on X.

Instead, I looked at an alternative reference [here](https://doi.org/10.1007/s42102-020-00036-9) and implemented Eq. (24) of that reference. In this reference, there is a dependence on R on both the RHS and LHS, so it makes more sense to me.

Please note that this equation also resembles the peridynamics partial stress tensor from [here](https://msp.org/jomms/2015/10-5/p05.xhtml)

BTW: this stress tensor is in general nonsymmetric (like the classical Piola stress tensor, see [here](https://doi.org/10.1016/j.jmps.2007.08.004)). So, it seems we would need to compute the full tensor, not only the 6 entries. I only included the 6 entries for now.

pabloseleson commented:
@streeve : Please take a look at the current implementation.

streeve commented on src/force/CabanaPD_Force_PMB.hpp:
```diff
@@ -161,6 +164,25 @@ class Force<ExecutionSpace, ForceModel<PMB, Elastic, ModelParams...>>
+            double fz_i = coeff * rz / r;
+
+            // sigma_xx
+            stress( i, 0 ) += 0.5 * fx_i * rx * vol( j );
```
```suggestion
            const double coeff = 0.5 * vol( j );
            stress( i, 0 ) += coeff * fx_i * rx;
```

streeve commented on src/force/CabanaPD_Force_PMB.hpp:
```diff
@@ -274,14 +325,15 @@ class Force<MemorySpace, ForceModel<PMB, Fracture, ModelParams...>>
               class MuView, class ParallelType>
     double computeEnergyFull( WType& W, const PosType& x, const PosType& u,
-                              DamageType& phi, const ParticleType& particles,
+                              DamageType& phi, ParticleType& particles,
```
Need to revert these energy function changes

streeve reviewed (commented):
Couple small things to clean up, but this is close to ready. @kinanbezem7 let me know if it makes sense how to fix the failing test or remaining comments or we can look at it in person

streeve commented on src/force/CabanaPD_Force_PMB.hpp:
```diff
@@ -321,6 +372,65 @@ class Force<MemorySpace, ForceModel<PMB, Fracture, ModelParams...>>
     }
+
+    template <class ParticleType, class MuView, class ParallelType>
+    double computeStressFull( ParticleType& particles, MuView& mu,
```
Should be `void`

kinanbezem7 replied:
Changed to void

streeve commented on src/force/CabanaPD_Force_PMB.hpp:
```diff
@@ -397,7 +507,7 @@ class Force<MemorySpace, ForceModel<LinearPMB, Elastic, ModelParams...>>
     template <class PosType, class WType, class ParticleType,
               class ParallelType>
     double computeEnergyFull( WType& W, const PosType& x, const PosType& u,
```
Missing linear PMB

streeve replied:
@kinanbezem7 Last changes needed are linear PMB + LPS stresses that match PMB

kinanbezem7 replied:
I'm trying to run the crack_branching problem just to check that stress is working. but i don't see it in the SILO output files. Does it not fall under OtherFields here in particles.hpp? 

```
#ifdef Cabana_ENABLE_SILO
        Cabana::Grid::Experimental::SiloParticleOutput::
            writePartialRangeTimeStep(
                "particles", local_grid->globalGrid(), output_step, output_time,
                0, n_local, getPosition( use_reference ), sliceForce(),
                sliceDisplacement(), sliceVelocity(),
                std::forward<OtherFields>( other )... );
```

streeve replied:
You've reminded me of a Silo-specific issue with matrix fields. As I recall they just don't show up in paraview. If you open an issue on your hdf5 problems I can try to help there.

streeve reviewed (changes requested):
Missing LPS stresses

streeve commented:
@pabloseleson can you take a look at the new example - the crack isn't centered. @kinanbezem7 there are a number of new files that look unintentional

kinanbezem7 commented:
> @pabloseleson can you take a look at the new example - the crack isn't centered. @kinanbezem7 there are a number of new files that look unintentional

Apologies, I got CabanaPD working on WSL on windows and then tried to work from there for this last part, so probably a bunch of files got thrown in. Still getting used to Github

streeve commented:
> > @pabloseleson can you take a look at the new example - the crack isn't centered. @kinanbezem7 there are a number of new files that look unintentional
> 
> Apologies, I got CabanaPD working on WSL on windows and then tried to work from there for this last part, so probably a bunch of files got thrown in. Still getting used to Github

No worries, but avoiding`git add .` will help. This branch should work again after the merge earlier today

kinanbezem7 commented:
@streeve Is there anything else that you need me to do with this?

streeve commented:
> @streeve Is there anything else that you need me to do with this?

I had hoped @pabloseleson would have time to take a look. Instead, you could remove the unintended files, remove the new example, and run the crack branching case with stress output posted here

kinanbezem7 commented:
> > @streeve Is there anything else that you need me to do with this?
> 
> I had hoped @pabloseleson would have time to take a look. Instead, you could remove the unintended files, remove the new example, and run the crack branching case with stress output posted here

I noticed there was a mistake with the LinearLPS stress  which I corrected and removed the extra files. I tested this for both LPS and PMB and looks like we are getting reasonable output, but I haven't looked into a proper validation of the magnitude of the stresses

streeve commented:
@kinanbezem7 I updated this branch on main and combined all commits. Can you test with the crack branching and post a result? Should be ready to merge if it looks good

streeve commented:
> Sure thing I'll try and get it done ASAP. Apologies things have been really hectic in the lab.

Hey no worries, other things getting merged simplified this a lot anyway

kinanbezem7 commented:
@streeve Just leaving a note here, I tried to run the crackbranching example with the elastic solver which failed due to the prenotch only being available with the fracture solver.
Instead I'm just setting the fracture energy to be really high so the prenotch still works but to avoid the crack progression so we can clearly see the stresses under load.

streeve commented:
> Just leaving a note here, I tried to run the crackbranching example with the elastic solver which failed due to the prenotch only being available with the fracture solver. Instead I'm just setting the fracture energy to be really high so the prenotch still works but to avoid the crack progression so we can clearly see the stresses under load.

Sounds good. Yeah, that's because currently in order to use a prenotch you need to store the full bond array and check it at each step. We could implement a non-fracture version with sparse bonds, but I'm not sure if anyone truly needs it at the moment

kinanbezem7 commented:
I tried to match the paper the best I could and here are the results I got. (I had to play around with the legend limits and I did not normalise the results like they do in the paper, but the patterns should be similar)
![yy](https://github.com/user-attachments/assets/6bba0040-6069-43c0-ae17-d60cb72e392d)
yy
![xy](https://github.com/user-attachments/assets/3063c1c2-7398-499e-aa94-cc1846eaaca1)
xy
![xx](https://github.com/user-attachments/assets/19d8fe0e-b9f9-40d5-9b30-c1c35b5f9503)
xx
![Screenshot 2025-01-24 at 04-24-20 Peridynamic stress is the static first Piola–Kirchhoff Virial stress - 1-s2 0-S0020768322000427-main pdf](https://github.com/user-attachments/assets/68830beb-bedc-4951-9a3e-968c4de1a96d)
![Screenshot 2025-01-24 at 04-23-50 Peridynamic stress is the static first Piola–Kirchhoff Virial stress - 1-s2 0-S0020768322000427-main pdf](https://github.com/user-attachments/assets/04fb215b-1050-4778-b29a-8d7dc837be93)

streeve reviewed (approved):
@pabloseleson I think this is ready. Add any further review you think is needed

pabloseleson commented:
@streeve Great to see progress on this! I am very sorry for the delay in reviewing. I plan to get to this soon.

streeve commented:
@pabloseleson this is increasingly likely to cause conflicts - if you can review with or without additional testing that would be best

pabloseleson commented on src/force/CabanaPD_Force_LPS.hpp:
```diff
@@ -245,6 +246,59 @@ class Force<MemorySpace, ForceModel<LPS, Elastic, NoFracture>>
+
+            double coeff = model.forceCoeff( s, xi, vol( j ), m( i ), m( j ),
+                                             theta( i ), theta( j ) );
+            coeff *= vol( j );
```
I believe a half factor is needed here like for the PMB: 
coeff *= 0.5 * vol( j );

pabloseleson commented on src/force/CabanaPD_Force_LPS.hpp:
```diff
@@ -502,6 +557,73 @@ class Force<MemorySpace, ForceModel<LPS, Elastic, Fracture>>
+                    double coeff =
+                        model.forceCoeff( s, xi, vol( j ), m( i ), m( j ),
+                                          theta( i ), theta( j ) );
+                    coeff *= vol( j );
```
I believe a half factor is needed here like for the PMB: 
coeff *= 0.5 * vol( j );

pabloseleson commented on src/force/CabanaPD_Force_LPS.hpp:
```diff
@@ -621,6 +744,61 @@ class Force<MemorySpace, ForceModel<LinearLPS, Elastic, NoFracture>>
+
+            double coeff = model.forceCoeff( linear_s, xi, vol( j ), m( i ),
+                                             m( j ), theta( i ), theta( j ) );
+            coeff *= vol( j );
```
I believe a half factor is needed here like for the PMB: 
coeff *= 0.5 * vol( j );

pabloseleson commented on src/force/CabanaPD_Force_PMB.hpp:
```diff
@@ -447,6 +562,61 @@ class Force<MemorySpace,
+            model.thermalStretch( linear_s, i, j );
+
+            // Linear PMB specific coefficient
+            double coeff = 0.5 * model.forceCoeff( linear_s, vol( j ) );
```
I would move the 0.5 factor to the next line for consistency with the other calculations, i.e.,:
            double coeff = model.forceCoeff( linear_s, vol( j ) );
            coeff *= 0.5 * vol( j );

streeve commented:
@pabloseleson can you make the changes you suggest?

pabloseleson commented on src/force/CabanaPD_Force_PMB.hpp:
```diff
@@ -447,6 +562,61 @@ class Force<MemorySpace,
+            double coeff = 0.5 * model.forceCoeff( linear_s, vol( j ) );
+            coeff *= vol( j );
+
+            const double fx_i = coeff * rx / r;
```
Since this is for the linear PMB, we should have instead:
 const double fx_i = coeff * xi_x / xi;
 const double fy_i = coeff * xi_y / xi;
 const double fz_i = coeff * xi_z / xi;

pabloseleson commented:
@streeve changes made

pabloseleson commented:
@streeve We can add a unit test for the stress calculation. This will be analogous to the unit test for the strain energy density. In the case of an isotropic extension with constant stretch \bar{s}, the diagonal stress tensor components are:

![Stress](https://github.com/user-attachments/assets/1e36d1ab-2c2f-4c7d-a5b3-477ef56eb72f)

and the off-diagonal are 0, where k is the bulk modulus.

pabloseleson commented on src/force/CabanaPD_Force_LPS.hpp:
```diff
@@ -245,6 +246,59 @@ class Force<MemorySpace, ForceModel<LPS, Elastic, NoFracture>>
+            const double fz_i = coeff * rz / r;
+
+            // Update stress tensor components
+            stress( i, 0, 0 ) += fx_i * rx;
```
In Lines 280-291, the stress calculation should use xi_x, xi_y, xi_z (like for the linearized version):

            stress( i, 0, 0 ) += fx_i * xi_x;
            stress( i, 1, 1 ) += fy_i * xi_y;
            stress( i, 2, 2 ) += fz_i * xi_z;

            stress( i, 0, 1 ) += fx_i * xi_y;
            stress( i, 1, 0 ) += fy_i * xi_x;

            stress( i, 0, 2 ) += fx_i * xi_z;
            stress( i, 2, 0 ) += fz_i * xi_x;

            stress( i, 1, 2 ) += fy_i * xi_z;
            stress( i, 2, 1 ) += fz_i * xi_y;

streeve replied:
Fixed

pabloseleson commented on src/force/CabanaPD_Force_LPS.hpp:
```diff
@@ -502,6 +557,73 @@ class Force<MemorySpace, ForceModel<LPS, Elastic, Fracture>>
+                    const double fz_i = muij * coeff * rz / r;
+
+                    // Update stress tensor components
+                    stress( i, 0, 0 ) += fx_i * rx;
```
Same here: 

In Lines 280-291, the stress calculation should use xi_x, xi_y, xi_z (like for the linearized version):

            stress( i, 0, 0 ) += fx_i * xi_x;
            stress( i, 1, 1 ) += fy_i * xi_y;
            stress( i, 2, 2 ) += fz_i * xi_z;

            stress( i, 0, 1 ) += fx_i * xi_y;
            stress( i, 1, 0 ) += fy_i * xi_x;

            stress( i, 0, 2 ) += fx_i * xi_z;
            stress( i, 2, 0 ) += fz_i * xi_x;

            stress( i, 1, 2 ) += fy_i * xi_z;
            stress( i, 2, 1 ) += fz_i * xi_y;

pabloseleson commented on src/force/CabanaPD_Force_PMB.hpp:
```diff
@@ -176,6 +177,57 @@ class Force<MemorySpace, ForceModel<PMB, Elastic, NoFracture, ModelParams...>>
+            const double fy_i = coeff * ry / r;
+            const double fz_i = coeff * rz / r;
+
+            stress( i, 0, 0 ) += fx_i * rx;
```
Same here:

In Lines 280-291, the stress calculation should use xi_x, xi_y, xi_z (like for the linearized version):

            stress( i, 0, 0 ) += fx_i * xi_x;
            stress( i, 1, 1 ) += fy_i * xi_y;
            stress( i, 2, 2 ) += fz_i * xi_z;

            stress( i, 0, 1 ) += fx_i * xi_y;
            stress( i, 1, 0 ) += fy_i * xi_x;

            stress( i, 0, 2 ) += fx_i * xi_z;
            stress( i, 2, 0 ) += fz_i * xi_x;

            stress( i, 1, 2 ) += fy_i * xi_z;
            stress( i, 2, 1 ) += fz_i * xi_y;

pabloseleson commented on src/force/CabanaPD_Force_PMB.hpp:
```diff
@@ -336,6 +389,67 @@ class Force<MemorySpace, ForceModel<PMB, Elastic, Fracture, ModelParams...>>
+                const double fy_i = coeff * ry / r;
+                const double fz_i = coeff * rz / r;
+
+                stress( i, 0, 0 ) += fx_i * rx;
```
Same here: 

In Lines 280-291, the stress calculation should use xi_x, xi_y, xi_z (like for the linearized version):

            stress( i, 0, 0 ) += fx_i * xi_x;
            stress( i, 1, 1 ) += fy_i * xi_y;
            stress( i, 2, 2 ) += fz_i * xi_z;

            stress( i, 0, 1 ) += fx_i * xi_y;
            stress( i, 1, 0 ) += fy_i * xi_x;

            stress( i, 0, 2 ) += fx_i * xi_z;
            stress( i, 2, 0 ) += fz_i * xi_x;

            stress( i, 1, 2 ) += fy_i * xi_z;
            stress( i, 2, 1 ) += fz_i * xi_y;

pabloseleson commented on src/force/CabanaPD_PMB.hpp:
```diff
@@ -342,6 +396,70 @@ class Force<MemorySpace,
+                model.thermalStretch( s, i, j );
+
+                double coeff =
+                    mu( i, n ) * model.forceCoeff( i, n, s, vol( j ) );
```
Should "n" be "j" in "model.forceCoeff", i.e., 

model.forceCoeff( i, j, s, vol( j ) )

?

pabloseleson commented on src/force/CabanaPD_PMB.hpp:
```diff
@@ -453,6 +571,60 @@ class Force<MemorySpace,
+            model.thermalStretch( linear_s, i, j );
+
+            // Linear PMB specific coefficient
+            double coeff = model.forceCoeff( i, n, linear_s, vol( j ) );
```
Should "n" be "j" in "model.forceCoeff", i.e., 

model.forceCoeff( i, j, linear_s, vol( j ) )

?

pabloseleson commented on examples/mechanics/plate_with_hole.cpp:
```diff
@@ -0,0 +1,164 @@
+        return true;
+    };
+
+    auto particles = CabanaPD::createParticles<memory_space, model_type>(
```
@streeve Could you please add the stress output here?

streeve replied:
The interface has changed. I will update

streeve replied:
Added

streeve commented on examples/mechanics/inputs/crack_branching_stress_test.json:
Why add this file if you're adding a new example?

streeve commented on src/force/CabanaPD_PMB.hpp:
```diff
@@ -270,7 +324,9 @@ class Force<MemorySpace,
                 else if ( mu( i, n ) > 0 )
                 {
-                    const double coeff = model.forceCoeff( i, n, s, vol( j ) );
+                    // const double coeff = model.forceCoeff( i, n, s, vol( j )
```
Do not leave commented code

streeve replied:
Fixed

streeve commented on src/force/CabanaPD_PMB.hpp:
```diff
@@ -257,7 +311,7 @@ class Force<MemorySpace,
+                getDistance( x, u, i, j, xi, r, s, rx, ry, rz );
 
                 model.thermalStretch( s, i, j );
 
```
You have addressed this

streeve commented:
@pabloseleson This has been updated to reflect other recent changes and so that the stress test compiles. The magnitude of the stress is wrong, but constant

pabloseleson commented on src/CabanaPD_Force.hpp:
```diff
@@ -91,19 +91,30 @@ getDistanceComponents( const PosType& x, const PosType& u, const int i,
-    const PosType& x, const PosType& u, const int i, const int j, double& xi,
-    double& s, double& xi_x, double& xi_y, double& xi_z )
+KOKKOS_INLINE_FUNCTION void
+getLinearizedDistance( const PosType& x, const PosType& u, const int i,
```
@streeve We could use "linear_s" instead of "s" in "double& s" for consistency with the function call. Same in other instances.

streeve replied:
I think it's implied within `getLinearizedDistance`

pabloseleson replied:
Note the expression in Line 127 is for the linearized stretch:

s = ( xi_x * eta_u + xi_y * eta_v + xi_z * eta_w ) / ( xi * xi );

However, if you consider it is implied, then it's probably OK.

pabloseleson commented on unit_test/tstForce.hpp:
```diff
@@ -482,11 +587,22 @@ void checkResults( HostParticleType aosoa_host, double local_min[3],
+
+            std::array<double, 3> f = { f_host( p, 0 ), f_host( p, 1 ),
+                                        f_host( p, 2 ) };
+            // Components xx, yy, zz, xy, xz, yz, yx, zx, zy
```
@streeve The order of components when we compute is:

stress( i, 0, 0 ) += fx_i * xi_x;
stress( i, 1, 1 ) += fy_i * xi_y;
stress( i, 2, 2 ) += fz_i * xi_z;
stress( i, 0, 1 ) += fx_i * xi_y;
stress( i, 1, 0 ) += fy_i * xi_x;
stress( i, 0, 2 ) += fx_i * xi_z;
stress( i, 2, 0 ) += fz_i * xi_x;
stress( i, 1, 2 ) += fy_i * xi_z;
stress( i, 2, 1 ) += fz_i * xi_y;

Should we keep the same order here?

streeve replied:
No, they're completely separate. And storing them this way makes the test easier

pabloseleson commented on src/force/CabanaPD_LPS.hpp:
```diff
@@ -245,6 +246,60 @@ class Force<MemorySpace, ForceModel<LPS, Elastic, NoFracture>>
+
+            double coeff = model.forceCoeff( s, xi, vol( j ), m( i ), m( j ),
+                                             theta( i ), theta( j ) );
+            coeff *= 0.5;
```
@streeve : Note before we had coeff *= 0.5 * vol( j ). However, forceCoeff already accounts for vol(j), so we were multiplying by vol(j) twice. We had this issue both for the PMB and LPS stress tensor computations. 

This is fixed now.

pabloseleson commented on unit_test/tstForce.hpp:
```diff
@@ -254,7 +297,65 @@ double computeReferenceNeighbors( const double delta, const int m )
+                {
+                    // We assume the dilatation and weighted volume are constant
+                    f_mag =
+                        ( model.theta_coeff * 2.0 * theta / weighted_volume +
```
@streeve I fixed the calculation of the reference LPS stress tensor by removing factor: ( 1.0 / num_neighbors ). 

This was needed in the energy because the dilatation was not under the integral, but not here.

streeve commented:
@pabloseleson you have not approved
